### ISSUE TITLE
In-band telemetry (MRI exercise)

### ISSUE THREAD

**Issue by suniljha555** [p4lang/tutorials#355]
Problem in keeping estimate of queue size in bytes:
Assuming each packet passes through both egress and ingress pipeline, my implementation is as follows
`register<bit<32>>(1) queue_size_bytes;`
Ingress pipeline code
`bit<32> temp;`
 `queue_size_bytes.read(temp,0);`
 `queue_size_bytes.write(0,(standard_metadata.packet_length+temp));`
Egress pipeline code
`bit<32> temp;`
 `queue_size_bytes.read(temp,0);`
 `queue_size_bytes.write(0,(temp-standard_metadata.packet_length));`

The link between S1 and S2 is shared by h1,h2, h11,h22. So when i send packets using send.py file which sends packets every one second then queue length is zero which is understandable as flow rate is very low but when iperf is started between h11 h22, the queue gets build and when iperf timer is finished queue_depth becomes zero but the queue size in bytes i calculated stucks at what the last value is.
Can someone please explain the reason? (and does iperf packet also passes through parser, ingress,egress pipeline,checksum and deparser?)
**Note:**
In first image q_depth represents `queue_size_bytes`  whereas in second image `swid` represents `queue_size_bytes` variable value .
![image](https://user-images.githubusercontent.com/16577683/81509397-062feb00-9328-11ea-83a1-c745188ed0ac.png)
![image](https://user-images.githubusercontent.com/16577683/81510016-750f4300-932c-11ea-8bd7-a8c8249f4487.png)
Is there any working demo of estimating/calculating queue size in bytes in v1model switch using P4?
Thank You

**Comment by jafingerhut** [comment]
There is one factor I had forgotten when suggesting an approach like this: after ingress processing is complete, I believe that it is possible that the v1model architecture can drop a packet without enqueuing it, and such dropped packets will never have egress processing performed for them.

Thus with the approach you are using, their size in bytes would never be subtracted from the register's contents.

I can imagine a P4 program that periodically calculates how many such packets were discarded, but one question for you is: is it worth going to that trouble?

In other words, what is your goal?  If your goal is to access queue lengths in bytes or "cells" in a real switch, then I would say look up the documentation of the real switch you are writing a P4 program for, and see if it already has this feature built in, unlike v1model which does not already have it built in.

If your goal is to do it by hook or by crook in the v1model architecture, why is that so important to you?

**Comment by suniljha555** [comment]
> There is one factor I had forgotten when suggesting an approach like this: after ingress processing is complete, I believe that it is possible that the v1model architecture can drop a packet without enqueuing it, and such dropped packets will never have egress processing performed for them.
> 
> Thus with the approach you are using, their size in bytes would never be subtracted from the register's contents.
> 
> I can imagine a P4 program that periodically calculates how many such packets were discarded, but one question for you is: is it worth going to that trouble?
> 
> In other words, what is your goal? If your goal is to access queue lengths in bytes or "cells" in a real switch, then I would say look up the documentation of the real switch you are writing a P4 program for, and see if it already has this feature built in, unlike v1model which does not already have it built in.
> 
> If your goal is to do it by hook or by crook in the v1model architecture, why is that so important to you?

I thought of a load balancing strategy to employ in data center topology based on queue size(bytes) (it's for my college project). No i don't have any real switch to work with, I am only working with simulators.. Can you suggest any other open switch architecture which provides such data about queues, if any?(i read in stanford slides sume_switch provides queue size in cells where cell size is 32 bytes)

yes you were right, 45% packets was dropped.(it dropped packets because of queue overflow because when i increased bottleneck bandwidth by 4 times only 0.22% packets were dropped and queue never build up )
![image](https://user-images.githubusercontent.com/16577683/81527526-2091b500-9378-11ea-9a4f-1b75df4071b6.png)
![image](https://user-images.githubusercontent.com/16577683/81529473-8a13c280-937c-11ea-8b5a-d1ad98b48caf.png)
![image](https://user-images.githubusercontent.com/16577683/81529502-9566ee00-937c-11ea-9c81-09f2bfb9fe9b.png)



Thank you very much for help

**Comment by jafingerhut** [comment]
The simple_switch program seems to me likely to cause you some trouble, if your goal is to simulate the performance of any strategies that can build up queues and have congested links between switches.

You can limit the rate that packets are scheduled from a single queue to a switch port with some configuration commands, but the current implementation only allows you to configure the maximum packet rate, not the maximum bit rate, so if you have mixes of different packet sizes, there is no way I can think of to emulate a real switch with a constant bit rate link.

In the absence of using those packet rate limits per queue, I believe simple_switch will schedule and do egress processing on packets waiting in a queue for a single output port as fast as your CPU core(s) will allow, and sometimes do so in a very bursty fashion, e.g. at an effective rate of gigabits per second for a short time, then a longer time of rate 0.  That is not what a real switch does (unless there are features like Ethernet pause flow control in use).

Maybe others besides me have found ways around these properties of simple_switch, or I was missing some kind of configuration, but in general I would be cautious about recording lots of performance related data, e.g. time stamps when packets began egress processing using standard_metadata.egress_global_timestamp, and look at that recorded data carefully to see if it is operating performance-wise in a way that suits the purposes of your experiments.

**Comment by suniljha555** [comment]
> The simple_switch program seems to me likely to cause you some trouble, if your goal is to simulate the performance of any strategies that can build up queues and have congested links between switches.
> 
> You can limit the rate that packets are scheduled from a single queue to a switch port with some configuration commands, but the current implementation only allows you to configure the maximum packet rate, not the maximum bit rate, so if you have mixes of different packet sizes, there is no way I can think of to emulate a real switch with a constant bit rate link.
> 
> In the absence of using those packet rate limits per queue, I believe simple_switch will schedule and do egress processing on packets waiting in a queue for a single output port as fast as your CPU core(s) will allow, and sometimes do so in a very bursty fashion, e.g. at an effective rate of gigabits per second for a short time, then a longer time of rate 0. That is not what a real switch does (unless there are features like Ethernet pause flow control in use).
> 
> Maybe others besides me have found ways around these properties of simple_switch, or I was missing some kind of configuration, but in general I would be cautious about recording lots of performance related data, e.g. time stamps when packets began egress processing using standard_metadata.egress_global_timestamp, and look at that recorded data carefully to see if it is operating performance-wise in a way that suits the purposes of your experiments.

Thanks for suggestions. I think i need to change metrics for my load balancing strategy. I have few more questions 
1.The metrics mentioned in INT documentation depends on switches architecture i.e what it provides in it's standard metadata?(because INT documentation mentions many metrics which are not available in PSA and v1model or do we have calculate some of the metrics ?)
2. Is there any working INT examples for v1model switch using P4-16 language?(i.e, using bitamp to collect metrics using INT headers.I  found few but all were faulty.Please share repository link if there is any) (I am working on load balancing for UDP flows using INT )

P4 tutorials are great . They are explained properly and clears most of the basic doubts. But I would like to suggest that if you can add one more tutorial where switch involves in both way communication i.e switch sends packets to controller and then controller pushes rules to it using P4runtime (i think present P4runtime exercise just pushes rules to the switch). It would be great, as i think this would complete the tutorials where one can find most basic things at one place including communication with controllers.(or any other repository link pointing to the above implementation)
Thanks. You've been a great help

**Comment by jafingerhut** [comment]
I do not know whether there are any public INT implementations or not.  I have re-asked on an issue here, which you can follow along any answers to, if you are interested: https://github.com/p4lang/p4-applications/issues/39#issuecomment-627039504

When creating an implementation of at least part of INT on the v1model architecture, it would of course be easiest if you only implemented the subset of INT that the v1model architecture directly supports.

Another option would be to enhance the C++ code that implements the v1model architecture in simple_switch so that measurements you wanted, were provided by the architecture itself.  For example, see this issue that suggests adding queue depths in units of bytes to simple_switch: https://github.com/p4lang/behavioral-model/issues/841

Since the C++ source code implementation of simple_switch is open source, in principle anyone in the world with the right skills can enhance it for themselves to add those missing measurements.  The tricky part is finding someone with the right skills who is interested in doing so, or finding someone interested who is able to gain the needed skills to do it.

"3. Can we run different switch programs in different switches in topology on simulators?"  I am not familiar enough with the code and scripts used to start multi-switch mininet setups in this repository, to know if it makes it straightforward to load different P4 programs into different switches.  In principle every emulated switch is just a simple_switch process with whatever command line arguments you want, and which compiled P4 program to run can be a command line argument when starting simple_switch, so as long as you can figure out how to make the program loaded in each one be what you want, it should be possible.

**Comment by suniljha555** [comment]
Thank you very much for help and suggestions.
